File: csfle/dotnet/aws/reader/CSFLE/InsertEncryptedDocument.cs
using System;
using System.Collections.Generic;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Encryption;
using Credentials;

namespace Insert
{

    class InsertEncryptedDocument
    {

        public static void Insert()
        {
            var credentials = new YourCredentials().GetCredentials();
            var connectionString = credentials["MONGODB_URI"];
            // start-key-vault
            var keyVaultNamespace = CollectionNamespace.FromFullName("encryption.__keyVault");
            // end-key-vault
            var coll = "patients";
            var db = "medicalRecords";
            var dbNamespace = $"{db}.{coll}";

            // start-kmsproviders
            var kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();
            var provider = "aws";
            var awsAccessKey = credentials["AWS_ACCESS_KEY_ID"];
            var awsSecretAccessKey = credentials["AWS_SECRET_ACCESS_KEY"];
            var awsKmsOptions = new Dictionary<string, object>
            {
               { "accessKeyId", awsAccessKey },
               { "secretAccessKey", awsSecretAccessKey }
            };
            kmsProviders.Add(provider, awsKmsOptions);
            // end-kmsproviders


            // start-schema
            var schema = new BsonDocument {
                { "bsonType", "object" },
                { "encryptMetadata", new BsonDocument{
                        { "keyId", "/key-id"}
                    }
                },
                { "properties", new BsonDocument {
                        { "ssn", new BsonDocument {
                                { "encrypt", new BsonDocument {
                                    { "bsonType", "int" },
                                    { "algorithm", "AEAD_AES_256_CBC_HMAC_SHA_512-Random"}
                                    }
                                }
                            }
                        },
                        { "bloodType", new BsonDocument {
                            { "encrypt" ,new BsonDocument {
                                    { "bsonType", "string" },
                                    { "algorithm", "AEAD_AES_256_CBC_HMAC_SHA_512-Random"}
                                    }
                                }
                            }
                        },
                        { "medicalRecords", new BsonDocument {
                                {"encrypt", new BsonDocument {
                                    { "bsonType", "array" },
                                    { "algorithm", "AEAD_AES_256_CBC_HMAC_SHA_512-Random"}
                                    }
                                }
                            }
                        },
                        { "insurance", new BsonDocument {
                                { "bsonType", "object" },
                                { "properties", new BsonDocument {
                                        { "policyNumber", new BsonDocument {
                                                { "encrypt", new BsonDocument {
                                                        { "bsonType", "int" },
                                                        { "algorithm", "AEAD_AES_256_CBC_HMAC_SHA_512-Random"}
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            };
            var schemaMap = new Dictionary<string, BsonDocument>();
            schemaMap.Add(dbNamespace, schema);
            // end-schema


            // start-extra-options
            var extraOptions = new Dictionary<string, object>()
            {
                { "mongocryptdSpawnPath", credentials["MONGOCRYPTD_PATH"]},
            };
            // end-extra-options

            var regularClientSettings = MongoClientSettings.FromConnectionString(connectionString);
            var regularClient = new MongoClient(regularClientSettings);
            var regularCollection = regularClient.GetDatabase(db).GetCollection<BsonDocument>(coll);

            // start-client
            var clientSettings = MongoClientSettings.FromConnectionString(connectionString);
            var autoEncryptionOptions = new AutoEncryptionOptions(
                keyVaultNamespace: keyVaultNamespace,
                kmsProviders: kmsProviders,
                schemaMap: schemaMap,
                extraOptions: extraOptions
                );
            clientSettings.AutoEncryptionOptions = autoEncryptionOptions;
            var secureClient = new MongoClient(clientSettings);
            // end-client

            // start-insert
            var sampleDocFields = new BsonDocument
            {
                { "name", "Jon Doe" },
                { "ssn", 145014000 },
                { "key-id", "demo-data-key"},
                { "bloodType", "AB-" },
                {
                    "medicalRecords", new BsonArray
                    {
                        new BsonDocument("weight", 180),
                        new BsonDocument("bloodPressure", "120/80")
                    }
                },
                {
                    "insurance", new BsonDocument
                    {
                        { "policyNumber", 123142 },
                        { "provider", "MaestCare" }
                    }
                }
            };

            // Construct an auto-encrypting client
            var secureCollection = secureClient.GetDatabase(db).GetCollection<BsonDocument>(coll);

            // Insert a document into the collection
            secureCollection.InsertOne(sampleDocFields);
            // end-insert
            // start-find
            Console.WriteLine("Finding a document with regular (non-encrypted) client.");
            var filter = Builders<BsonDocument>.Filter.Eq("name", "Jon Doe");
            var regularResult = regularCollection.Find(filter).Limit(1).ToList()[0];
            Console.WriteLine($"\n{regularResult}\n");

            Console.WriteLine("Finding a document with encrypted client, searching on an encrypted field");
            var secureResult = secureCollection.Find(filter).Limit(1).First();
            Console.WriteLine($"\n{secureResult}\n");
            // end-find
        }
    }
}

File: csfle/node/gcp/reader/insert_encrypted_document.js
const mongodb = require("mongodb");
const { ClientEncryption } = require("mongodb-client-encryption");
const { MongoClient, Binary } = mongodb;

const { getCredentials } = require("./your_values");
credentials = getCredentials();

var db = "medicalRecords";
var coll = "patients";
var namespace = `${db}.${coll}`;
// start-kmsproviders
const kmsProviders = {
  gcp: {
    email: credentials["GCP_EMAIL"],
    privateKey: credentials["GCP_PRIVATE_KEY"],
  },
};
// end-kmsproviders

const connectionString = credentials.MONGODB_URI;

// start-key-vault
const keyVaultNamespace = "encryption.__keyVault";
// end-key-vault

// start-schema
const schema = {
  bsonType: "object",
  encryptMetadata: {
    keyId: "/key-id",
  },
  properties: {
    insurance: {
      bsonType: "object",
      properties: {
        policyNumber: {
          encrypt: {
            bsonType: "int",
            algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Random",
          },
        },
      },
    },
    medicalRecords: {
      encrypt: {
        bsonType: "array",
        algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Random",
      },
    },
    bloodType: {
      encrypt: {
        bsonType: "string",
        algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Random",
      },
    },
    ssn: {
      encrypt: {
        bsonType: "int",
        algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Random",
      },
    },
  },
};

var patientSchema = {};
patientSchema[namespace] = schema;
// end-schema

// start-extra-options
const extraOptions = {
  mongocryptdSpawnPath: credentials["MONGOCRYPTD_PATH"],
};
// end-extra-options

// start-client
const secureClient = new MongoClient(connectionString, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
  autoEncryption: {
    keyVaultNamespace,
    kmsProviders,
    schemaMap: patientSchema,
    extraOptions: extraOptions,
  },
});
// end-client
const regularClient = new MongoClient(connectionString, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
});

async function main() {
  try {
    await regularClient.connect();
    try {
      await secureClient.connect();
      // start-insert
      try {
        const writeResult = await secureClient
          .db(db)
          .collection(coll)
          .insertOne({
            name: "Jon Doe",
            ssn: 241014209,
            bloodType: "AB+",
            "key-id": "demo-data-key",
            medicalRecords: [{ weight: 180, bloodPressure: "120/80" }],
            insurance: {
              policyNumber: 123142,
              provider: "MaestCare",
            },
          });
      } catch (writeError) {
        console.error("writeError occurred:", writeError);
      }
      // end-insert
      // start-find
      console.log("Finding a document with regular (non-encrypted) client.");
      console.log(
        await regularClient.db(db).collection(coll).findOne({ name: /Jon/ })
      );

      console.log(
        "Finding a document with encrypted client, searching on an encrypted field"
      );
      console.log(
        await secureClient.db(db).collection(coll).findOne({ name: /Jon/ })
      );
      // end-find
    } finally {
      await secureClient.close();
    }
  } finally {
    await regularClient.close();
  }
}
main();
